### PR TITLE
improve handling of changelog processing for backports

### PR DESCRIPTION
**Pull request by nickfyson**
This PR fixes handling of backporting the changelog when the only listed change for a version is not applicable for the older targeted release.

### Merge / deployment checklist

- [x] Confirm this change is backwards compatible with existing workflows.
- [x] Confirm the [readme](https://github.com/github/codeql-action/blob/main/README.md) has been updated if necessary.
- [x] Confirm the [changelog](https://github.com/github/codeql-action/blob/main/CHANGELOG.md) has been updated if necessary.

### CODE REVIEW

File: .github/update-release-branch.py [hunk 1]
```diff
@@ -174,6 +175,58 @@ def get_today_string():
   today = datetime.datetime.today()
   return '{:%d %b %Y}'.format(today)
 
+def process_changelog_for_backports(source_branch_major_version, target_branch_major_version):
+
+  # changelog entries use a speficic format to indicate
```
Comment (henrymercer) [inline .github/update-release-branch.py:180]:
```suggestion
  # changelog entries can use the following format to indicate
```

File: .github/update-release-branch.py [hunk 2]
```diff
@@ -174,6 +175,58 @@ def get_today_string():
...
+  # that they only apply to newer versions
+  regex = re.compile(r'\[v(\d+)\+ only\]')
```
Comment (henrymercer) [inline .github/update-release-branch.py:182]: Rename to `some_versions_only_regex`?

File: .github/update-release-branch.py [hunk 3]
```diff
@@ -174,6 +175,58 @@ def get_today_string():
...
+        if int(target_branch_major_version) < int(match.group(1)):
+          continue
+
+      if line.startswith('## '):
+        line = line.replace(f'## {source_branch_major_version}', f'## {target_branch_major_version}')
+        if found_content == False:
+          # we have found two headings in a row, so we need to add the placeholder message.
+          output += 'No user facing changes.\n'
+        found_content = False
+        output += '\n' +line + '\n\n'
```
Comment (henrymercer) [inline .github/update-release-branch.py:219]:
```suggestion
        output += f'\n{line}\n\n'
```

**Review by henrymercer — DISMISSED** [review]
Looks good.  Another approach would be parsing the document in full into a version -> changelog notes map, performing our tweaks on the map, then rewriting to Markdown.  I think that might be more readable, but it's probably not better enough to justify changing it — what do you think?

File: .github/update-release-branch.py (hunk 2 above)
Comment (nickfyson) [reply .github/update-release-branch.py]: 👍

**Review by henrymercer — DISMISSED** [review]

**Comment by nickfyson** [comment]
Yeah, I started doing that and then decided it was overkill. But then had to work pretty hard to make this approach even vaguely grokkable. 😅

I think I'll leave it with this approach, and if/when it gets more complicated then refactor it 'properly'. 😄

File: .github/update-release-branch.py [hunk 4]
```diff
@@ -174,6 +175,58 @@ def get_today_string():
...
+
+  output = ''
+
+  with open('CHANGELOG.md', 'r') as f:
+
+    # until we find the first section, just duplicate all lines
+    while True:
+      line = f.readline().rstrip()
+
+      output += line + '\n'
```
Comment (aeisenberg) [inline .github/update-release-branch.py:192]: This is minor and non-blocking, but why `rstrip` and then add `\n`? If the original changelog has trailing whitespace (even though it shouldn't), then would it be best to keep the new one as close to the original as possible?

File: .github/update-release-branch.py [hunk 5]
```diff
@@ -174,6 +175,58 @@ def get_today_string():
...
+      if line.startswith('## '):
```
Comment (aeisenberg) [inline .github/update-release-branch.py:193]: We _shouldn't_ have to handle this, but this script will fail if there is no `## ` line. Maybe just to be safe include a fallback to handle this case (I could imagine this _might_ happen if we're testing something).

**Review by aeisenberg — APPROVED** [review]
Looks reasonable to me. Two non-blocking suggestions.

File: .github/update-release-branch.py (hunk 5 above)
Comment (nickfyson) [reply .github/update-release-branch.py:195]: I've added an explicit error message in this situation, to try and avoid confusion should it arise. 👍

**Review by aeisenberg — APPROVED** [review]

**Pull request merged**